flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,463 @@
+    }
+    0
+  }
+
+  override def putNormalizedKey(record: Row, target: MemorySegment, offset: Int, numBytes: Int)
+      : Unit = {
+    var bytesLeft = numBytes
+    var currentOffset = offset
+
+    var i = 0
+    var j = 0
+    while (i < numLeadingNormalizableKeys && bytesLeft > 0) {
+      var len = normalizedKeyLengths(i)
+      len = if (bytesLeft >= len) len else bytesLeft
+
+      val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+      val element = record.productElement(keyPositions(i))
+      if (len > 0) {
+        // write a null byte with padding
+        if (element == null) {
+          target.putBoolean(currentOffset, true)
+          // write padding
+          j = 0
+          while (j < len - 1) {
+            target.put(currentOffset + j, 0.toByte)
+            j += 1
+          }
+        }
+        // write a non-null byte with key
+        else {
+          target.putBoolean(currentOffset, false)
+          // write key
+          comparator.putNormalizedKey(element, target, currentOffset, len - 1)
+        }
+      }
+
+      bytesLeft -= len
+      currentOffset += len
+      i += 1
+    }
+  }
+
+  override def extractKeys(record: Any, target: Array[AnyRef], index: Int): Int = {
+    val len = comparators.length
+    var localIndex = index
+    var i = 0
+    while (i < len) {
+      val element = record.asInstanceOf[Row].productElement(keyPositions(i))
+      localIndex += comparators(i).extractKeys(element, target, localIndex)
+      i += 1
+    }
+    localIndex - index
+  }
+
+}
+
+object RowComparator {
+  private class NullChecker extends Comparable[AnyRef] with Serializable {
+    override def compareTo(obj: AnyRef): Int =
+      throw new UnsupportedOperationException("This should never be called.")
+    override def equals(obj: Any): Boolean =
+      obj == null
+  }
+
+  /**
+   * @return creates auxiliary fields for normalized key support
+   */
+  private def createAuxiliaryFields(
+      comparators: Array[TypeComparator[_]])
+    : (List[Int], Int, Int) = {
+    comparators.takeWhile {
+      c =>
+        // As long as the leading keys support normalized keys, we can build up the composite key.
+        // The first comparator decides whether we need to invert the key direction.
+        // If a successor does not agree on the inversion direction, it cannot be part of the
+        // normalized key.
+        c.supportsNormalizedKey() &&
+          c.invertNormalizedKey() == comparators(0).invertNormalizedKey()
+    }.foldLeft((List[Int](), 0, 0)) {
+      case ((normKeyLengths, numLeadingNormKeys, normKeyPrefixLen), comparator) =>
+        val len = comparator.getNormalizeKeyLen
+
+        if (len < 0) {
+          throw new RuntimeException("Comparator " + comparator.getClass.getName +
+            " specifies an invalid length for the normalized key: " + len)
+        }
+
+        val newNormKeyLengths = (len + 1) :: normKeyLengths // add one for a null byte
```

Comment:
`len` can be `Integer.MAX_VALUE` (as for example for `StringComparator`). Adding 1 results in an overflow.